org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/ZigBeeThingTypeMatcher.java

```diff
@@ -0,0 +1,145 @@
+ * <ul>
+ * <li>The matcher requires a match of all properties in the thing-type.
+ * <li>The properties to be used for matching can contain extra properties that aren't used. This allows the user to
+ * pass in all known properties of the device, and the matcher will use whatever properties are in the thing-type.
+ * </ul>
+ * <p>
+ * The property file is a simple text file with the thingtypeUID at the beginning of the line, followed by a set of
+ * comma separate attributes.
+ * eg.
+ * <ul>
+ * <li>philips_sml001,vendor=Philips,modelId=SML001
+ * <li>smartthings_motionv4,vendor=SmartThings,modelId=motionv4
+ * </ul>
+ *
+ * @author Chris Jackson - Initial Implementation
+ *
+ */
+public class ZigBeeThingTypeMatcher {
+    private final Logger logger = LoggerFactory.getLogger(ZigBeeThingTypeMatcher.class);
+
+    private final Map<String, List<RequiredProperty>> discoveryProperties = new HashMap<>();
+
+    private final static String DISCOVERY_PROPERTIES_FILE = "/discovery.txt";
+
+    /**
+     * Matches a set of properties to a single thing type. If no match is found, null is returned.
+     * The matcher checks all registered thing types. If multiple matches are found, null is returned.
+     *
+     * @param properties a Map of strings containing the properties to match
+     * @return the {@link ThingTypeUID} of the matched thing type, or null if no single thing type matched the filter
+     *         properties
+     */
+    public ThingTypeUID matchThingType(Map<String, String> properties) {
+        // Only create the index once
+        synchronized (discoveryProperties) {
+            if (discoveryProperties.isEmpty()) {
+                try {
+                    readDiscoveryProperties();
+                } catch (IOException exception) {
+                    logger.debug("IOException reading ZigBee discovery properties", exception);
+                    return null;
+                }
+            }
+        }
+
+        String matchedThingType = null;
+
+        for (Entry<String, List<RequiredProperty>> discoveryThing : discoveryProperties.entrySet()) {
+
+            if (discoveryThing.getValue().stream().allMatch(
+                    requiredProperty -> requiredProperty.value.equals(properties.get(requiredProperty.name)))) {
+                if (matchedThingType != null) {
+                    // Don't allow duplicate matches
+                    logger.debug("Duplicate match {} and {}", matchedThingType, discoveryThing.getKey());
+                    return null;
+                }
+                matchedThingType = discoveryThing.getKey();
+            }
+        }
+
+        if (matchedThingType == null) {
+            return null;
+        }
+
+        return new ThingTypeUID(ZigBeeBindingConstants.BINDING_ID, matchedThingType);
+    }
+
+    private void readDiscoveryProperties() throws IOException {
+        InputStream input = ZigBeeThingTypeMatcher.class.getResourceAsStream(DISCOVERY_PROPERTIES_FILE);
```

Comment:
Where is this stream ever closed again? Looks like a resource leak.

Reply:
Good point - I'll fix this.